### DISCUSSION TITLE
Request : Add session ID on Hook Environnement variable

### DISCUSSION THREAD

**Discussion by guedz** [drakkan/sftpgo#640] [Ideas]
Hello,

Is it possible to add a "Session ID" and/or "Connection ID" to hook environnement variable to permit to consolidate the data that I generate when a user opens several sessions at the same time (for example)? 

Best regards and again thanks for your work/help

**Comment by drakkan** [comment]
https://github.com/drakkan/sftpgo/commit/a7c0b07a2ab88d63cbc4a382e46d50ff1e3bd391

**Comment by guedz45** [comment]
Hello,

Thanks

[06-12-2021 10:14:51][SSH][demo][172.20.0.1] allow connection
[06-12-2021 10:14:51][SSH][demo][172.20.0.1] Status login: OK (1)
[06-12-2021 10:14:55][SFTP][SFTP_250070a0c1161b3c6f9c166c883110cdd6f7d1e4428666463ab0870eddb1b702_1][demo][172.20.0.1] mkdir the folder /in - Status: OK
[06-12-2021 10:14:58][SFTP][SFTP_250070a0c1161b3c6f9c166c883110cdd6f7d1e4428666463ab0870eddb1b702_1][demo][172.20.0.1] mkdir the folder /out - Status: OK
[06-12-2021 10:16:49][SFTP][demo][172.20.0.1] disconnected. Duration: 118645 milliseconds

Is it possible to retrieve this SESSION ID also in post_login_hook, pre_login_hook and post_disconnect_hook ?

Best regards

**Reply by drakkan** [reply]
Hi,

- pre_login_hook is executed before a session id is created, currently the session id is created after the user logins
- post_login_hook is executed after the user log in, but SSH is a multiplexing protocol, after the authentication the client can open one or more SFTP sessions, an SCP session or execute an SSH command, for each session we create a separate session id otherwise if you want to terminate only one SFTP session this would be not possible.

So adding something like this will require a discrete refactoring and some assumptions for the hook receiver, I don't plan to work on this sorry.

**Comment by guedz45** [comment]
No problem i understand.
I thought that "session id" was an internal counter to sftpgo generated at each opening of session because we find this information well in the JSON.
I think I can manage with this (if it is the same information) 

Example : SFTPGO_LOGIND_USER => connection_id

{"level":"debug","time":"2021-12-06T10:14:51.328","sender":"osfs","connection_id":"250070a0c1161b3c6f9c166c883110cdd6f7d1e4428666463ab0870eddb1b702","message":"root directory \"/srv/sftpgo/data/homedir/demo\" for user \"demo\" does not exist, try to create, mkdir error: <nil>"}
{"level":"info","time":"2021-12-06T10:14:51.328","sender":"SSH","connection_id":"250070a0c1161b3c6f9c166c883110cdd6f7d1e4428666463ab0870eddb1b702","message":"User \"demo\" logged in with \"password\", from ip \"172.20.0.1\", client version \"SSH-2.0-FileZilla_3.56.2\""}

The difference seems to be:

- add protocol before
- add _{number incremented i suppose) at the end

**Comment by guedz45** [comment]
After a quick test, just add "connection_id" that i can see on the primary log on JSON SFTPGO_LOGIND_USER 
for me that would be more than enough 

UPDATE: I will even say more on reflection just getting the connection_id from pre-login to disconnect-login through events would be sufficient.
